drivers/gpio/gpio_sam.c

```diff
@@ -22,64 +22,109 @@ struct gpio_sam_config {
+#define DEV_DATA(dev) \
+	((struct gpio_sam_runtime *const)(dev)->driver_data)
 
-static int gpio_sam_config_pin(Pio * const pio, u32_t mask, int flags)
+static int gpio_sam_config_pin(struct device *dev, Pio * const pio, u32_t mask,
+			       int flags)
 {
+	struct gpio_sam_runtime *const dev_data = DEV_DATA(dev);
+
+	if (!(flags & (GPIO_OUTPUT | GPIO_INPUT))) {
+		/* Neither input nor output mode is selected */
+
+		/* Disable the interrupt. */
+		pio->PIO_IDR = mask;
+		/* Disable pull-up. */
+		pio->PIO_PUDR = mask;
+		/* Disable pull-down. */
+		pio->PIO_PPDDR = mask;
+		/* Let the PIO control the pin (instead of a peripheral). */
+		pio->PIO_PER = mask;
+		/* Disable output. */
+		pio->PIO_ODR = mask;
+
+		return 0;
+	}
+
 	/* Setup the pin direcion. */
-	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
+	if (flags & GPIO_OUTPUT) {
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			/* Set the pin. */
+			pio->PIO_SODR = mask;
+		}
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			/* Clear the pin. */
+			pio->PIO_CODR = mask;
+		}
 		pio->PIO_OER = mask;
 	} else {
 		pio->PIO_ODR = mask;
 	}
 
+	/* Note: Input is always enabled. */
+
+	if (flags & GPIO_ACTIVE_LOW) {
+		dev_data->invert |= mask;
+	} else {
+		dev_data->invert &= ~mask;
+	}
+
 	/* Setup interrupt configuration. */
-	if (flags & GPIO_INT) {
-		if (flags & GPIO_INT_DOUBLE_EDGE) {
-			return -ENOTSUP;
-		}
 
-		/* Enable the interrupt. */
-		pio->PIO_IER = mask;
+	/* Disable the interrupt. */
+	pio->PIO_IDR = mask;
+	/* Disable additional interrupt modes. */
+	pio->PIO_AIMDR = mask;
+
+	if ((flags & GPIO_INT) &&
+	    (flags & (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1))) {
+		if (!((flags & GPIO_INT_LOW_0) && (flags & GPIO_INT_HIGH_1))) {
+			/* Enable additional interrupt modes. */
+			pio->PIO_AIMER = mask;
+
+			if (flags & GPIO_INT_EDGE) {
+				pio->PIO_ESR = mask;
+			} else {
+				pio->PIO_LSR = mask;
+			}
 
-		/* Enable the additional interrupt modes. */
-		pio->PIO_AIMER = mask;
+			bool rising_edge = flags & GPIO_INT_HIGH_1;
 
-		if (flags & GPIO_INT_EDGE) {
-			pio->PIO_ESR = mask;
-		} else {
-			pio->PIO_LSR = mask;
-		}
+			if ((flags & GPIO_INT_LEVELS_LOGICAL) &&
+			    (flags & GPIO_ACTIVE_LOW)) {
+				rising_edge = !rising_edge;
+			}
 
-		if (flags & GPIO_INT_ACTIVE_HIGH) {
-			/* Set to high-level or rising edge. */
-			pio->PIO_REHLSR = mask;
-		} else {
-			/* Set to low-level or falling edge. */
-			pio->PIO_FELLSR = mask;
+			if (rising_edge) {
+				/* Set to high-level or rising edge. */
+				pio->PIO_REHLSR = mask;
+			} else {
+				/* Set to low-level or falling edge. */
+				pio->PIO_FELLSR = mask;
+			}
 		}
-	} else {
-		/* Disable the interrupt. */
-		pio->PIO_IDR = mask;
 
-		/* Disable ther additional interrupt modes. */
-		pio->PIO_AIMDR = mask;
+		/* Enable the interrupt. */
+		pio->PIO_IER = mask;
 	}
 
 	/* Setup Pull-up resistor. */
-	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	if (flags & GPIO_PULL_UP) {
```

Comment:
Does it really make sense to let the user turn on both the pull up and pull down resistors? Sorry if I am not understanding the idea, but it doesn't seem sensible.

I guess part of the Zephyr philosophy is "minimal error checking". But maybe we could add some asserts for inconsistent behavior like this if you agree it doesn't make sense, but don't want to add code or runtime cost to check this sort of thing in production builds.

Reply:
I don't think there is any real life use case where an application may want to enable both pull up and pull down resistors even if some hardware will happily do it. I don't perform checks not to add extra run time code but indeed we could handle this with an assert. I'll add it in include/gpio.h file.